torch/distributed/rpc/backend_registry.py

```diff
@@ -265,6 +265,27 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
         # process initializes its PyTorch CUDA states.
         torch.cuda.init()
 
+        # Check local devices in device_maps and devices are all valid.
+        local_devices = set(rpc_backend_options.devices) if rpc_backend_options.devices else set()
+        for worker_name in rpc_backend_options.device_maps:
+            local_devices.update(rpc_backend_options.device_maps[worker_name].keys())
+
+        if len(local_devices) > 0 and not all([
+            max(local_devices) < torch.cuda.device_count(),
+            min(local_devices) >= 0,
+        ]):
+            raise ValueError(
+                f"Invalid device in TensorPipe options on {name}:\n"
+                f"device_maps = {rpc_backend_options.device_maps},\n"
+                f"devices = {rpc_backend_options.devices}"
+            )
```

Comment:
The reason that these two checks are in separate places was because `_tensorpipe_check_device_maps` is after the construction of `TensorPipeAgent`. And we need to make sure the devices we pass to the agent ctor is correct, as subsequent RPCs in `all_gather` will already start using those devices information. Let refactor these two checks, such that we check local devices before ctor and remote ones after.